### PR TITLE
Prototype: Server-side block attributes sourcing

### PR DESCRIPTION
**Pull request by aduth**
This pull request seeks to explore an approach to block attributes sourcing on the server. In other words, it seeks to account for attributes whose values are derived from HTML (or post meta). It should be considered a prototype, but it currently implements most all of the current source supports.

Example:

```bash
curl 'http://localhost:8889/wp-json/wp/v2/posts/95?context=edit&_fields=content'  -H 'X-WP-Nonce: [...]' -H 'Cookie: wordpress_logged_in_[...]=[...]' | jq .
```

```json
{
  "content": {
    "raw": "<!-- wp:paragraph {\"align\":\"center\",\"className\":\"my-custom-class\"} -->\n<p class=\"has-text-align-center my-custom-class\">Hello world</p>\n<!-- /wp:paragraph -->\n\n<!-- wp:image {\"id\":20,\"sizeSlug\":\"large\"} -->\n<figure class=\"wp-block-image size-large\"><img src=\"http://localhost:8889/wp-content/uploads/2019/11/stars-1024x681.jpeg\" alt=\"\" class=\"wp-image-20\"/><figcaption>Caption!</figcaption></figure>\n<!-- /wp:image -->",
    "rendered": "\n<p class=\"has-text-align-center my-custom-class\">Hello world</p>\n\n\n\n<figure class=\"wp-block-image size-large\"><img src=\"http://localhost:8889/wp-content/uploads/2019/11/stars-1024x681.jpeg\" alt=\"\" class=\"wp-image-20\" srcset=\"http://localhost:8889/wp-content/uploads/2019/11/stars-1024x681.jpeg 1024w, http://localhost:8889/wp-content/uploads/2019/11/stars-300x199.jpeg 300w, http://localhost:8889/wp-content/uploads/2019/11/stars-768x510.jpeg 768w, http://localhost:8889/wp-content/uploads/2019/11/stars-1536x1021.jpeg 1536w, http://localhost:8889/wp-content/uploads/2019/11/stars-2048x1361.jpeg 2048w\" sizes=\"(max-width: 1024px) 100vw, 1024px\" /><figcaption>Caption!</figcaption></figure>\n",
    "protected": false,
    "block_version": 1,
    "blocks": [
      {
        "name": "core/paragraph",
        "attributes": {
          "align": "center",
          "className": "my-custom-class",
          "content": "Hello world"
        },
        "inner_blocks": []
      },
      {
        "name": "core/image",
        "attributes": {
          "id": 20,
          "sizeSlug": "large",
          "url": "http://localhost:8889/wp-content/uploads/2019/11/stars-1024x681.jpeg",
          "alt": "",
          "caption": "Caption!",
          "title": null,
          "href": null,
          "rel": null,
          "linkClass": null,
          "linkTarget": null
        },
        "inner_blocks": []
      }
    ]
  },
}
```

**Implementation Notes:**

The parsing relies on [`DOMDocument`](https://www.php.net/manual/en/class.domdocument.php), querying using [`DOMXPath`](https://www.php.net/manual/en/class.domxpath.php) by first converting the block type attribute selector to an equivalent XPath selector using a bundled, modified version of a third-party library [PHP Selector](https://github.com/tj/php-selector).

In an effort to best demonstrate its usage, additional changes include:

- A mechanism for server-registering all Gutenberg `block.json` manifests
- Include a new `content.blocks` field on the REST API posts responses
- Defining default-supported attributes for blocks registered on the server (`className`, `align`, `anchor`)

**Open Questions:**

For me, the main questions and concerns surrounding this approach include:

- Is it performant enough?
- Are `DOMDocument` and the PHP Selector utilities resilient enough, and do they account for modern HTML and CSS?
   - There may be other options to explore here, including [`symfony/css-selector`](https://github.com/symfony/css-selector)
- What permissions would we require for access to this raw data?

### CODE REVIEW

**Review by mcsf — COMMENTED** [review]
Impressive!

> Is it performant enough?
> Are DOMDocument and the PHP Selector utilities resilient enough, and do they account for modern HTML and CSS?

Performance is the first thing one wonders about. To answer both those questions, there's probably nothing like getting this in the hands of testers. To that effect: we've been relatively liberal at pushing experimental client-side interfaces and know how to approach things there; how could we do the same here?

**Comment by mcsf** [comment]
Resurfacing #7342, as it's in the same domain and something we could easily tackle now.

**Comment by aduth** [comment]
>To that effect: we've been relatively liberal at pushing experimental client-side interfaces and know how to approach things there; how could we do the same here?

I think a safe route (albeit perhaps more labor-intensive) would be to first consider extracting some of the "additional" prerequisite work that I had to bundle into this pull request:

>In an effort to best demonstrate its usage, additional changes include:
>
>- A mechanism for server-registering all Gutenberg `block.json` manifests
>- Include a new `content.blocks` field on the REST API posts responses
>- Defining default-supported attributes for blocks registered on the server (`className`, `align`, `anchor`)

The first and last of these would complement ongoing work in [Trac#47620](https://core.trac.wordpress.org/ticket/47620) (cc @gziolo, @spacedmonkey), since while the endpoint proposed there would expose registered blocks, it's currently the case that very few core blocks are actually registered on the server.

It would be good to have some feedback from REST API folks as well, at least so far as how we expose this data through that interface. It might be a task worth considering separate from how the attributes actually become populated during the parse. I can plan to bring it up in their weekly meeting this upcoming Thursday.

Lastly, while the plugin was a nice venue for quick prototyping, and we may be able to merge it in some experimental form, moving forward we might want to develop in Trac, or at least develop there some additional hooks necessary for a more solid solution. That this implementation replaces the default parser may not be how we ultimately want to go about doing this (e.g. perhaps we want a more "pure" HTML parser result from `parse_blocks`, and a separate `parse_blocks_with_sourced_attributes` that considers external factors like block registries, "current post", etc). Maybe @dmsnell has some thoughts here.

One specific task which otherwise limits the effectiveness of this in the plugin is a means to filter block registration settings on the server. Currently this is not possible, and in this implementation I manually apply this by re-registering the core set of blocks.

File: lib/class-wp-sourced-attributes-block-parser.php [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+	 */
+	function get_html_sourced_attribute( $block, $attribute_schema ) {
+		$document = new DOMDocument();
+		try {
+			// loadHTML may log warnings for unexpected markup.
+			// phpcs:ignore
+			@$document->loadHTML( '<html><body>' . $block['innerHTML'] . '</body></html>' );
+		} catch ( Exception $e ) {
+			return null;
+		}
```
Comment (dmsnell) [inline lib/class-wp-sourced-attributes-block-parser.php]: when I have worked with `DOMDocument` I found it extremely helpful to extract the setup code into a separate function that abstracts away the noisy quirks.

```php
$document = parseHTML( $block['innerHTML'] );
if ( null === $document ) {
	return null;
}
```

it's a small thing to abstract but I find in my experience it worth it especially as we learn about the settings we need to activate with whitespace and with parse-handling.

File: lib/parser.php [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+/**
+ * Filters to override default block parser with Gutenberg's custom
+ * implementation.
+ *
+ * @since 6.9.0
+ */
+function gutenberg_replace_block_parser_class() {
+	return 'WP_Sourced_Attributes_Block_Parser';
+}
+add_filter( 'block_parser_class', 'gutenberg_replace_block_parser_class' );
```
Comment (dmsnell) [inline lib/parser.php]: this is how the system was designed to work, though I believe that you should be able to skip creating the helper function.

```php
add_filter( 'block_parser_class', 'WP_Sourced_Attributes_Block_Parser' );
```

**Comment by dmsnell** [comment]
Thanks for getting back to server-side parsing of attributes @aduth.

This change seems to mitigate the problem of having attributes sourced from HTML and I think it's good to get it in Core. If we end up thinking this is the way to go I'd only want to consider eventually merging it into the default parser, though I'd prefer to find a way to do that which doesn't mash all that code into the default parser.

From our discussions I know that you are aware of the fact that this only addresses one of a few concerns with sourced attributes, that the desire to have all attributes available to a parser with no knowledge of the block implementations is still unresolved by this change.

It seems though that we have been pushing the limit of what sourced attributes are offering and some people are really wishing we didn't have as much of them; I think that in some ways the biggest problem we're addressing is one of quantity and not of quality. If we can surface the attributes for the Core blocks then most people will be happy.

This patch gets those attributes available to the PHP on the running server while leaving them inaccessible to any other parser. That's better than leaving them unavailable to every parser. 🙂

File: lib/blocks.php [hunk 3]
```diff
@@ -27,6 +27,39 @@ function gutenberg_get_registered_social_link_blocks() {
...
 
+/**
+ * Registers blocks using block manifests discovered in block library.
+ *
+ * @since 6.9.0
+ */
+function gutenberg_register_block_types() {
+	$registry = WP_Block_Type_Registry::get_instance();
+
+	$block_manifests = glob( dirname( dirname( __FILE__ ) ) . '/packages/block-library/src/*/block.json' );
```
Comment (gziolo) [inline lib/blocks.php]: I don't think we have `block.json` for all core blocks, those which are dynamic don't have this metadata file provided because we still didn't resolve the following issues:
- `supports` is implemented only on the client-side
- translations aren't tackled for fields defined in the JSON file, there is the proposal for the client-side https://github.com/WordPress/gutenberg/pull/16088, but we don't have anything for the server-side

Those aren't blockers for this proposal if we were to use only `attributes` though. So maybe it would be a good idea to move attributes to the `block.json` file to better promote this format.

File: gutenberg.php [hunk 4]
```diff
@@ -122,7 +122,7 @@ function gutenberg_pre_init() {
 	// Strip '-src' from the version string. Messes up version_compare().
 	$version = str_replace( '-src', '', $wp_version );
 
-	if ( version_compare( $version, '5.0.0', '<' ) ) {
+	if ( version_compare( $version, '5.2.0', '<' ) ) {
```
Comment (gziolo) [inline gutenberg.php]: The change to increase the minimal version of WordPress and PHP could be extracted as its own PR and landed soon.

Comment (Soean) [reply gutenberg.php]: @gziolo We already have #15809 which increases the minimum version to 5.2.

Comment (gziolo) [reply gutenberg.php]: Let's land it then 💯

**Comment by spacedmonkey** [comment]
Please consider this [ticket](https://core.trac.wordpress.org/ticket/48529) in core. I believe this has to land before we can continue this work. 

I also believe how `editor_script`, `script`, `editor_style` and `style` are handled in gutenberg and PHP are different. PHP using a handles where as javascript uses urls. There will need to be changed in PHP to handle url before this can be merged.

File: lib/blocks.php [hunk 5]
```diff
@@ -27,6 +27,39 @@ function gutenberg_get_registered_social_link_blocks() {
...
+	foreach ( $block_manifests as $block_manifest ) {
```
Comment (spacedmonkey) [inline lib/blocks.php]: Should some level of validation be happening here?

File: lib/class-wp-sourced-attributes-block-parser.php [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+	 * Returns a sourced attribute value for a block, for attribute type which
+	 * sources from HTML.
+	 *
+	 * @param WP_Block_Parser_Block $block            Parsed block object.
+	 * @param array                 $attribute_schema Attribute schema for
+	 *                                                individual attribute to
+	 *                                                be parsed.
+	 * @return mixed                                  Sourced attribute value.
+	 */
+	function get_html_sourced_attribute( $block, $attribute_schema ) {
```
Comment (spacedmonkey) [inline lib/class-wp-sourced-attributes-block-parser.php]: This call is going to be expensive from a compute level. Is there anyway we can cache the result?

File: lib/class-wp-sourced-attributes-block-parser.php [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+		$document = new DOMDocument();
```
Comment (spacedmonkey) [inline lib/class-wp-sourced-attributes-block-parser.php]: It seems like this class has some [requirements](https://www.php.net/manual/en/dom.requirements.php). Can we confirm that the libxml package is currently a required one for WP Core?

File: lib/parser.php (hunk 2 above)
Comment (spacedmonkey) [reply lib/parser.php]: This helper is useful, as it allows for this filter to be unhooked.

File: phpcs.xml.dist [hunk 8]
```diff
@@ -3,7 +3,7 @@
 	<description>Sniffs for WordPress plugins, with minor modifications for Gutenberg</description>
 
 	<rule ref="PHPCompatibility"/>
-	<config name="testVersion" value="5.2-"/>
+	<config name="testVersion" value="5.6-"/>
```
Comment (spacedmonkey) [inline phpcs.xml.dist]: This change should happen in another PR

File: lib/parser.php [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+ * Given a CSS selector string, returns an equivalent XPath selector.
+ *
+ * @access private
+ * @since 6.9.0
+ *
+ * @param string $selector CSS selector.
+ * @return string          Equivalent XPath selector.
+ */
+function _wp_css_selector_to_xpath( $selector ) {
+	/*
```
Comment (spacedmonkey) [inline lib/parser.php]: I would add a filter here, to high jack this bahviour

File: lib/parser.php [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+/**
+ * Given a CSS selector string, returns an equivalent XPath selector.
+ *
+ * @access private
+ * @since 6.9.0
+ *
+ * @param string $selector CSS selector.
+ * @return string          Equivalent XPath selector.
+ */
+function _wp_css_selector_to_xpath( $selector ) {
```
Comment (spacedmonkey) [inline lib/parser.php]: This function needs PHP unit tests.

File: lib/class-wp-sourced-attributes-block-parser.php [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+ */
+
+/**
+ * Class WP_Sourced_Attributes_Block_Parser
+ *
+ * Parses a document and constructs a list of parsed block objects
+ *
+ * @since 6.9.0
+ */
+class WP_Sourced_Attributes_Block_Parser extends WP_Block_Parser {
```
Comment (spacedmonkey) [inline lib/class-wp-sourced-attributes-block-parser.php]: Unit tests?

File: phpcs.xml.dist (hunk 8 above)
Comment (Soean) [reply phpcs.xml.dist]: We merged this in #15809

File: lib/class-wp-sourced-attributes-block-parser.php (hunk 1 above)
Comment (aduth) [reply lib/class-wp-sourced-attributes-block-parser.php]: > when I have worked with `DOMDocument` I found it extremely helpful to extract the setup code into a separate function that abstracts away the noisy quirks.

That seems like a reasonable revision, for sure! I expect it would also make writing the tests a little nicer, since I'd not need to lump all the error cases for this function otherwise intended specifically at sourcing values.

File: lib/parser.php (hunk 2 above)
Comment (aduth) [reply lib/parser.php]: >this is how the system was designed to work, though I believe that you should be able to skip creating the helper function.

I'd tend to agree with @spacedmonkey here, though it certainly piques my interest that this syntax can work 🤔

File: lib/blocks.php (hunk 3 above)
Comment (aduth) [reply lib/blocks.php]: > we still didn't resolve the following issues:

You'll have to forgive me since it's been a while that I've revisited some of those specific details of the JSON manifest. Working through this prototype forced me to consider how we _would_ implement at least some of those supports on the server (`className`, `align`, `anchor` are implemented in this pull request).

For translations, I seem to recall something about how we considered to wrap the translateable fields via `__` et. al., automatically? I'm not sure exactly how we determine the domain in that case.

As a prototype, I'm also fine to start splitting those off into their own individual tasks. It was at least interesting to explore the feasibility of pulling them in and highlighting some of these shortcomings (notably `supports`).

File: gutenberg.php (hunk 4 above)
Comment (aduth) [reply gutenberg.php]: > The change to increase the minimal version of WordPress and PHP could be extracted as its own PR and landed soon.

@gziolo @Soean Thanks, I'll make sure this is removed in a rebase.

For context, it was only needed due to lack of support of non-numeric values for [`array_reduce`](https://www.php.net/manual/en/function.array-reduce.php) third parameter. Not a big issue to refactor if it came to it, but I assumed (and supported by #15809!) that it would be fine to bump the minimum requirement.

**Comment by aduth** [comment]
> Please consider this [ticket](https://core.trac.wordpress.org/ticket/48529) in core. I believe this has to land before we can continue this work.

Thanks for the pointer! Those changes seem much-needed, I agree. For how it impacts this pull request, I think `$supports` support is likely the most pressing, in how it might impact what's currently implemented as `gutenberg_add_default_attributes`.

> I also believe how `editor_script`, `script`, `editor_style` and `style` are handled in gutenberg and PHP are different. PHP using a handles where as javascript uses urls. There will need to be changed in PHP to handle url before this can be merged.

Before this pull request is merged? Or the Trac ticket? I'm not really sure how those fields relate to the effort here.

File: lib/blocks.php (hunk 5 above)
Comment (aduth) [reply lib/blocks.php]: > Should some level of validation be happening here?

Do you mean that it has necessary properties to consider it a valid block manifest?

There are some simple checks below, both to account that the file could be parsed as JSON, and that it has a `name`. We could expand on this.

https://github.com/WordPress/gutenberg/blob/1e359a8d467b23c33c664f6c277c2549866be2dd/lib/blocks.php#L41

Comment (spacedmonkey) [reply lib/blocks.php]: Validation of types, like it string, int, array etc?

File: lib/class-wp-sourced-attributes-block-parser.php (hunk 6 above)
Comment (aduth) [reply lib/class-wp-sourced-attributes-block-parser.php]: > This call is going to be expensive from a compute level. Is there anyway we can cache the result?

I'd thought a bit about what might make sense to cache. Since the HTML of each block will likely be unique, I don't know that we would want to cache either the loaded HTML or the queried results, as the cache hit ratio would be very low.

What might make sense, depending on whether it makes a measurable difference:

- Caching the `$document` itself.
   - If constructing `DOMDocument` is expensive (I don't know that it is)
- Caching the converted XPath selectors
   - The conversion may or may not be expensive (might also depend which implementation we choose), but there's a higher likelihood we would reuse those on a a per-block-type basis (e.g. every paragraph will be running the `p` selector).

Comment (spacedmonkey) [reply lib/class-wp-sourced-attributes-block-parser.php]: How about cache the attributes, in say post meta?

File: lib/class-wp-sourced-attributes-block-parser.php (hunk 7 above)
Comment (aduth) [reply lib/class-wp-sourced-attributes-block-parser.php]: > It seems like this class has some [requirements](https://www.php.net/manual/en/dom.requirements.php). Can we confirm that the libxml package is currently a required one for WP Core?

From the page you link, it says "libxml is enabled by default".

We could still have some graceful fallback here for environments where it's explicitly disabled, although it would be unable to populate `$attributes`, yes.

File: lib/parser.php (hunk 9 above)
Comment (aduth) [reply lib/parser.php]: > I would add a filter here, to high jack this bahviour

Seems reasonable, sure 👍

File: lib/parser.php (hunk 10 above)
Comment (aduth) [reply lib/parser.php]: > This function needs PHP unit tests.

Yep, I have no plans to merge anything without sufficient tests.

File: lib/blocks.php (hunk 5 above)
Comment (aduth) [reply lib/blocks.php]: > Validation of types, like it string, int, array etc?

If I understand you correctly, we probably should want something like what exists in [`WP_Block_Type#prepare_attributes_for_render`](https://github.com/WordPress/wordpress-develop/blob/6cf77dbbb6d6b289d81dd2d33c742d150129b774/src/wp-includes/class-wp-block-type.php#L140-L157), though I expect this would be applied at the time that `$attributes` are being sourced.

File: lib/blocks.php (hunk 3 above)
Comment (gziolo) [reply lib/blocks.php]: > You'll have to forgive me since it's been a while that I've revisited some of those specific details of the JSON manifest. Working through this prototype forced me to consider how we _would_ implement at least some of those supports on the server (`className`, `align`, `anchor` are implemented in this pull request).

Nice, I missed that, sorry about it :(


> For translations, I seem to recall something about how we considered to wrap the translateable fields via `__` et. al., automatically? I'm not sure exactly how we determine the domain in that case.

There needs to be the `textDomain` field declared in the `block.json` file. You can check my prototype for JS side as a reference: https://github.com/WordPress/gutenberg/pull/16088.

I don't think it is a concern though in the context of `attributes`. I just wanted to raise awareness of that. The general agreement was that `attributes` shouldn't be translatable.

> As a prototype, I'm also fine to start splitting those off into their own individual tasks. It was at least interesting to explore the feasibility of pulling them in and highlighting some of these shortcomings (notably `supports`).

Yes, `supports` seems like the only place which can cause issues for the proposed code.

File: lib/class-wp-sourced-attributes-block-parser.php (hunk 6 above)
Comment (aduth) [reply lib/class-wp-sourced-attributes-block-parser.php]: > How about cache the attributes, in say post meta?

Hm, I'd have to think about it more, but that does seem like a good idea. In fact, it might then make sense to run this sourcing logic when a post is saved, rather than at parse-time (the parse would just read the cached result).

File: lib/class-wp-sourced-attributes-block-parser.php (hunk 7 above)
Comment (spacedmonkey) [reply lib/class-wp-sourced-attributes-block-parser.php]: We may need to change the requires for WP core.

**Comment by spacedmonkey** [comment]
In the [RFC](https://github.com/WordPress/gutenberg/blob/master/docs/rfc/block-registration.md#editor-script), editor_script, script, editor_style and style are defines as URLs. However PHP registered block use script / style handles. See [wp_enqueue_registered_block_scripts_and_styles](https://github.com/WordPress/WordPress/blob/master/wp-includes/script-loader.php#L2864-L2891). 
The `wp_enqueue_style` and `wp_enqueue_script` can not currently handle passing a url to them. 

TL:DR If you pass a URL in editor_script, script, editor_style and style fields, is not going to work and may break things.

> Before this pull request is merged? Or the Trac ticket? I'm not really sure how those fields relate to the effort here.

I think that [#48529](https://core.trac.wordpress.org/ticket/48529) has to land and we need to decide how the fields are handled in [#47620](https://core.trac.wordpress.org/ticket/47620) before we merge anything here.

**Comment by spacedmonkey** [comment]
The following blocks are missing block.json
tag-cloud
shortcode
search
rss
navigation-menu
legacy-widget
latest-posts
latest-comments
embed
categories
calendar
block
archives

See this [ticket](https://core.trac.wordpress.org/ticket/48529) and this [patch](https://core.trac.wordpress.org/attachment/ticket/48529/48529-implement.diff) that resolves the issue.

**Comment by gziolo** [comment]
> The following blocks are missing block.json

> See this ticket and this patch that resolves the issue.

Those files updated in the patch are replaced with the files from packages installed from npm. Your changes would be erased on the next run of `npm run build` or `npm run build:dev`. They need to be modified in Gutenberg. In addition, we need to land first the part which @aduth described in his comment https://github.com/WordPress/gutenberg/pull/18414#discussion_r346321150:

> Working through this prototype forced me to consider how we _would_ implement at least some of those supports on the server (`className`, `align`, `anchor` are implemented in this pull request).

We also didn't move translatable fields to `block.json` files because of:

> For translations, I seem to recall something about how we considered to wrap the translateable fields via __ et. al., automatically?

We don't have a code in place which would do it with PHP code when registering blocks from `block.json`.

**Comment by kadamwhite** [comment]
I'd like to propose exposing this block structure as a subresource, `post/:id/blocks`, instead of necessarily listing it in `content`; while it does represent post content, it's both more structural and in most cases not needed in the same contexts as the raw or rendered content. Making it a subresource would require an additional request or an `_embed` when this data is needed, but it would avoid somewhat duplicating the content we return in one response.

**Comment by aduth** [comment]
This was always meant to be a prototype exploration, so I'm going to close this as it's not in a mergeable state. It can be useful for future reference implementation.

As far as its current status, there were pending architectural revisions that ought to be explored in any future implementation:

- Caching https://github.com/WordPress/gutenberg/pull/18414#discussion_r345131348
- REST route https://github.com/WordPress/gutenberg/pull/18414#issuecomment-570275519

This pull request also included additional changes required by—but not directly related to—the addition of server-side attribute sourcing.

From the original comment:

>A mechanism for server-registering all Gutenberg block.json manifests
>Include a new content.blocks field on the REST API posts responses
>Defining default-supported attributes for blocks registered on the server (className, align, anchor)

Respective status of each:

- I believe @gziolo has his eyes on server-registering of `block.json`
- The effort here enables `content.blocks`, and so would be dependent on this implementation. In any case, given https://github.com/WordPress/gutenberg/pull/18414#issuecomment-570275519, it might not be encouraged to implement as part of the existing posts endpoint
- Depending if it is implemented first in Gutenberg, an initial step of server-side block `supports` is filterable block registration, proposed at [Trac#49615](https://core.trac.wordpress.org/ticket/49615).

**Comment by gziolo** [comment]
> * I believe @gziolo has his eyes on server-registering of `block.json`

Yes, the plan is to introduce new helper method that allows registering block using a new utility function that works with `block.json` as discussed in https://github.com/WordPress/gutenberg/pull/19786#issuecomment-579943306. In addition to that, we have WP-CLI changes tracked in https://github.com/wp-cli/scaffold-command/issues/141 to make it possible to include translatable strings in `block.json`.

File: lib/parser.php [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+ * @since 6.9.0
+ */
+function gutenberg_replace_block_parser_class() {
+	return 'WP_Sourced_Attributes_Block_Parser';
+}
+add_filter( 'block_parser_class', 'gutenberg_replace_block_parser_class' );
+
+/**
+ * Given a registered block type settings array, assigns default attributes.
+ * This must be called manually, as there is currently no way to hook to block
```
Comment (gziolo) [inline lib/parser.php:21]: Now, that `register_block_type_args` filter was introduced with https://core.trac.wordpress.org/ticket/49615, we can add this functionality in WordPress core. I will extract this function and propose it as an enhancement to the registration process on the server.

Comment (gziolo) [reply lib/parser.php:21]: The proposal is ready at https://github.com/WordPress/wordpress-develop/pull/383.
